### PR TITLE
Add recipe for napari

### PR DESCRIPTION
**Pull request by jni**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] ~~If static libraries are linked in, the license of the static library is packaged.~~
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by jni** [comment]
@conda-forge/help-python Do I need to add entry points even if they are defined in setup.py?

https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#python-entry-points

**Comment by jni** [comment]
Gah, looks like we didn't include the license file in the source distribution until now. Will get this fixed shortly.

File: recipes/napari/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  # If getting the source from GitHub, remove the line above,
+  # uncomment the line below, and modify as needed.
+  # url: https://github.com/napari/{{ name }}/archive/{{ version }}.tar.gz
```
Comment (scopatz) [inline recipes/napari/meta.yaml]: Please remove extra comment lines too

File: recipes/napari/meta.yaml [hunk 2]
```diff
@@ -13,7 +13,6 @@ source:
   sha256: 0e4a20f20be5db9224383e090d64718fb42c5359cdd8ee8be5c01076b7df230d
 
 build:
-  noarch: python
```
Comment (isuruf) [inline recipes/napari/meta.yaml]: You can keep this and have a constraint on python version instead of doing skipping.

Comment (jni) [reply recipes/napari/meta.yaml]: @isuruf great! It felt wrong to remove it. But could you point me to how to do it and where it is documented? The docs I saw all said that if it didn’t run on all the currently supported conda-forge Python versions, I should remove noarch. Happy to contribute an update to the docs if you show me how to update this correctly! Thanks!

File: recipes/napari/meta.yaml (hunk 1 above)
Comment (jni) [reply recipes/napari/meta.yaml]: Thanks! I kinda wanted to leave the ones that I thought I might refer to in the future, eg if we move to a GitHub-centric release process. Is it ok if I leave it around for future maintainers?

**Comment by jni** [comment]
@isuruf @scopatz thanks for the prompt review! Could you answer my question about entry points above? Ie should I add them even if they are defined in setup.py?

File: recipes/napari/meta.yaml (hunk 2 above)
Comment (isuruf) [reply recipes/napari/meta.yaml]: Docs at https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python is confusing. See https://github.com/conda-forge/awswrangler-feedstock/blob/440b2218dea38355a3314ff12c63c1daedd45e66/recipe/meta.yaml for an example.

**Comment by jni** [comment]
@conda-forge/help-python @conda-forge/pyopengl 

There appears to be an error installing PyOpenGL, which we depend on. Any idea why?

e.g.

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=84545&view=logs&jobId=10bda42a-d420-5c28-a43f-bc4e66405e5c&taskId=bef775d0-a1a2-55a3-9721-17147029baf0&lineStart=1616&lineEnd=1617&colStart=1&colEnd=1

Thank you!

**Comment by mcfletch** [comment]
The error messages suggests you're missing an OpenGL driver:
```
==> script output <==
stdout: Warning: Missing OpenGL driver, install with yum install mesa-libGL-devel or equivalent
```
if you *do* have an OpenGL driver, then for some reason we're not finding it. Potentially this might be a problem with using Wayland or the like?

**Comment by jni** [comment]
@mcfletch I don't understand how to fix this in the recipe or elsewhere, or why it affects this build but not the [vispy](https://github.com/conda-forge/vispy-feedstock)/[pyopengl](https://github.com/conda-forge/pyopengl-feedstock) builds. I don't see anything about opengl drivers in those recipes...?

**Comment by mcfletch** [comment]
(Should preface this with the fact that I don't know anything about conda-forge's process or patterns, I'm just an upstream developer).

PyOpenGL has been getting updates today and yesterday, so it's possible this is a bug attributed to something in PyOpenGL changing, but it *seems* the build is pulling the older beta, not the release candidate from yesterday.

The error log, as far as I read it, is saying that a "post-link" script failed. I don't know what the post-link script would be doing, but my guess from the error, is that it's trying to import something from PyOpenGL and e.g. run a smoke test or something of that nature.

**Comment by mcfletch** [comment]
So I see another project here: https://github.com/conda-forge/pyopengl-feedstock/issues/18 that seems to have encountered the same issue. I'm don't know enough about conda-forge to know whether the fix (adding the mesa lib) was "the right solution", but that was what the other issue seems to have been resolved with.

HTH,
Mike

File: recipes/napari/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    - numpy
+    - scipy
+    - scikit-image
+    - vispy
+    - dask
+    - zarr
+    - fsspec
+    - imageio
+    - ipykernel
+    - jupyter-client<6.0.0
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
    - jupyter-client <6.0.0
```

File: recipes/napari/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  # If getting the source from GitHub, remove the line above,
+  # uncomment the line below, and modify as needed.
+  # url: https://github.com/napari/{{ name }}/archive/{{ version }}.tar.gz
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
```

File: recipes/napari/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    - backcall
+    - pluggy
+    - PyOpenGl
+    - PySide2
+    - wrapt
+    - numpydoc
+
+test:
+  # Some packages might need a `test/commands` key to check CLI.
+  # List all the packages/modules that `run_test.py` imports.
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
```

File: recipes/napari/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    - zarr
+  commands:
+    - setx QT_API "pyside2"  # [win]
+    - export QT_API=pyside2  # [unix]
+    - pytest --pyargs napari
+
+about:
+  home: http://napari.org
+  license: BSD-3-Clause
+  # The license_family, i.e. "BSD" if license is "BSD-3-Clause". (optional)
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
```

File: recipes/napari/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  license_family: BSD
+  # It is strongly encouraged to include a license file in the package,
+  # (even if the license doesn't require it) using the license_file entry.
+  # See https://conda.pydata.org/docs/building/meta-yaml.html#license-file
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
```

File: recipes/napari/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  license_file: LICENSE
+  summary: 'a fast n-dimensional image viewer in Python'
+
+  # The remaining entries in this section are optional, but recommended.
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
```

File: recipes/pytest-ordering/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+requirements:
+  host:
+    - pip
+    - pytest
+    - python
+    - setuptools
+  run:
+    - pytest
+    - python
+    - setuptools
```
Comment (jakirkham) [inline recipes/pytest-ordering/meta.yaml]:
```suggestion
```

File: recipes/pytest-ordering/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+package:
+  name: "{{ name|lower }}"
+  version: "{{ version }}"
+
+source:
+  url: "https://github.com/ftobia/{{ name }}/archive/{{ version }}.tar.gz"
+  sha256: c1149011a3bec662d4dd92427197bd5d02546f380b3568eb6217efb743614309
+
+build:
+  number: 0
```
Comment (jakirkham) [inline recipes/pytest-ordering/meta.yaml]:
```suggestion
  number: 0
  noarch: python
```

File: recipes/napari/meta.yaml (hunk 1 above)
Comment (jakirkham) [reply recipes/napari/meta.yaml]: This was done.

File: recipes/napari/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  sha256: 56d41f596587b0877d31936972b8c53eab5e070e39377e1b97270d71f211793c
+
+build:
+  noarch: python
+  number: 0
+  script: "{{ PYTHON }} -m pip install . -vv"
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
  script: "{{ PYTHON }} -m pip install . -vv"
  entry_points:
    - napari = napari.__main__:main
```

File: recipes/napari/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    - numpy
+    - scipy
+    - scikit-image
+    - vispy
+    - dask
+    - zarr
+    - fsspec
+    - imageio
+    - ipykernel
+    - jupyter-client <6.0.0
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
    - jupyter_client <6.0.0
```

File: recipes/napari/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    - pytest-qt
+    - xarray
+    - pandas
+    - pytest-ordering
+    - pytest-timeout
+    - zarr
+  commands:
+    - setx QT_API "pyside2"  # [win]
+    - export QT_API=pyside2  # [unix]
+    - pytest --pyargs napari
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]: Let's go ahead and comment these for now. We can debug the segfault in the feedstock as it seems unrelated to Napari (as it is pure Python).

```suggestion
    #- pytest --pyargs napari
```

File: recipes/napari/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+test:
+  requires:
+    - pytest
+    - pytest-faulthandler
+    - pytest-qt
+    - xarray
+    - pandas
+    - pytest-ordering
+    - pytest-timeout
+    - zarr
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
    - zarr
  imports:
     - napari
```

File: recipes/napari/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  commands:
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
  commands:
    - napari --info
```

File: recipes/napari/meta.yaml [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    - xarray
+    - pandas
+    - pytest-ordering
+    - pytest-timeout
+    - zarr
+  imports:
+     - napari
+  commands:
+    - setx QT_API "pyside2"  # [win]
+    - export QT_API=pyside2  # [unix]
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
    - export QT_API=pyside2  # [unix]
    - napari --info
```

File: recipes/napari/meta.yaml [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    - napari --info
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
    - DISPLAY=localhost:1.0 xvfb-run -a bash -c 'napari --info'
```

File: recipes/napari/meta.yaml [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    - pandas
+    - pytest-ordering
+    - pytest-timeout
+    - zarr
+  imports:
+     - napari
+  commands:
+    - setx QT_API "pyside2"  # [win]
+    - export QT_API=pyside2  # [unix]
+    - DISPLAY=localhost:1.0 xvfb-run -a bash -c 'napari --info'
```
Comment (jakirkham) [inline recipes/napari/meta.yaml]:
```suggestion
    - which napari
    #- DISPLAY=localhost:1.0 xvfb-run -a bash -c 'napari --info'
```

**Comment by jakirkham** [comment]
Thanks @jni! 😄
